pantheon-bundle/src/main/java/com/redhat/pantheon/jcr/JcrResources.java

```diff
@@ -20,11 +23,24 @@ public static void rename(Resource resource, String newName) throws RepositoryEx
         session.move(resource.getPath(), resource.getParent().getPath() + "/" + newName);
     }
 
-    /*
-     * calculates a hash for a string
+    /**
+     * calculates a checksum hash for a string using the adler32 algorithm
      *
      */
     public static HashCode hash(String str) {
         return Hashing.adler32().hashString(str == null ? "" : str, Charsets.UTF_8);
     }
+
+    /**
+     * Calculates a checksum hash for the contents for an input stream usng the adler32 algorithm.
+     * @param is The input stream. The stream is NOT closed after reading from it.
+     * @return A {@link HashCode} for the contents of the given InputStream
+     * @throws IOException If there is a problem accessing the contents of the InputStream
+     */
+    public static HashCode hash(InputStream is) throws IOException {
+        try (HashingInputStream his = new HashingInputStream(Hashing.adler32(), is)) {
+            while (his.read() != -1) ;
```

Comment:
This is interesting. I suppose this is to load the entire stream into memory as a prerequisite to hash. I'm curious if this would necessitate a call to Stream.reset() later or something?

Reply:
The use of a `HashingInputStream` is to **avoid** loading the entire content into memory; it hashes as it consumes.

But you bring up a good point: the stream is potentially useless after it has been consumed. The way we are using it right now is safe as it is being closed, but a comment in the javadoc wouldn't hurt.

Reply:
Never mind... I did have a comment there already 😊

Reply:
> The use of a HashingInputStream is to avoid loading the entire content into memory; it hashes as it consumes.

Well, I understand that's the goal, but `while (his.read() != -1) ;` makes me think that that's not what's happening. The stream reads until there is nothing left, thereby loading everything into memory, *then* it hashes. Is that not what's happening?

Reply:
It does read the whole stream, but not into memory. It hashes as it reads, modifying the hash as it consumes the contents. The end result is the same as if the entire content were hashed.

Reply:
Ok, so you are asserting that the hashing actually happens in the `read()` method, and that the subsequent call to `hash()` doesn't actually do any work, it just returns the precomputed hash. If that's the case, then I'm good with it.

Reply:
correct. 
But this is not because of the HashingInputStream only, it also depends on the hashing function behind it, in this case adler32, which happens to operate in that way you described (confirmed by stepping through it).